include/seqan3/alignment/pairwise/execution/alignment_executor_two_way.hpp

```diff
@@ -44,36 +44,54 @@ namespace seqan3::detail
  * This alignment executor provides an additional buffer over the computed alignments to allow
  * a two-way execution flow. The alignment results can then be accessed in an order-preserving manner using the
  * alignment_executor_two_way::bump() member function.
+ *
+ * ### Bucket structure
+ *
+ * Since it is not clear how many results a single invocation of the given algorithm produces the buffered results
+ * are placed into buckets. The number of available buckets is determined by the execution policy. In sequential
+ * execution mode only one bucket is available and only one invocation is buffered at a time. In the parallel execution,
+ * a bucket is allocated for every element of the underlying resource.
  */
 template <std::ranges::viewable_range resource_t,
           typename alignment_algorithm_t,
           std::semiregular value_t,
           typename execution_handler_t = execution_handler_sequential>
 //!\cond
-    requires std::ranges::forward_range<resource_t> &&
-             std::copy_constructible<alignment_algorithm_t>
+    requires std::ranges::forward_range<resource_t> && std::copy_constructible<alignment_algorithm_t>
 //!\endcond
 class alignment_executor_two_way
 {
 private:
     /*!\name Resource types
      * \{
      */
-    //!\brief The underlying resource type augmented with an index and a chunked view.
-    using chunked_resource_type = typename chunked_indexed_sequence_pairs<resource_t>::type;
+    //!\brief The underlying resource type.
+    using resource_type = std::ranges::all_view<resource_t>;
     //!\brief The iterator over the underlying resource.
-    using chunked_resource_iterator = std::ranges::iterator_t<chunked_resource_type>;
+    using resource_iterator_type = std::ranges::iterator_t<resource_type>;
     //!\}
 
     /*!\name Buffer types
      * \{
      */
     //!\brief The internal buffer.
-    using buffer_type       = std::vector<value_t>;
+    using buffer_type = std::vector<value_t>;
+    //!\brief The type of the bucket container.
+    using bucket_type = std::vector<buffer_type>;
     //!\brief The pointer type of the buffer.
-    using buffer_pointer    = std::ranges::iterator_t<buffer_type>;
+    using buffer_iterator_type = std::ranges::iterator_t<buffer_type>;
+    //!\brief The iterator type over the bucket container.
+    using bucket_iterator_type = std::ranges::iterator_t<bucket_type>;
     //!\}
 
+    //!\brief Return status for seqan3::detail::alignment_executor_two_way::underflow.
+    enum underflow_status
+    {
+        non_empty_buffer, //!< The buffer is not fully consumed yet and contains at least one element.
```

Comment:
No a single bucket represents the results of one algorithm invocation. The buffer is the entire bucket list. So after calling underflwo I check if the buffer is empty or not. So if underflow calls x algorithms and any of them buffers a result no matter in which bucket it is not empty. So I don't need to call underflow again within the bump function. Because this function either returns a result or end of file/resource.

Reply:
> The buffer is the entire bucket list

ah ok. I think what confused me is that there is a `buffer_type` member variable which is the value type of `bucket_type`. That's why I thought a bucket contains a buffer. Maybe you can rename that type then?

Reply:
still open but not a blocker, maybe for future considerations

Reply:
yes, I wanted to rename anyway a lot of things. So the next step is to move it out of alignment into core and rename there some of the rather unspecific names.